File: server/game_server/graph.go
package main

import "fmt"

type NodeId int
type EdgeId int

type Graph struct {
	nodes map[NodeId]Node
	edges map[EdgeId]Edge
}

func NewGraph() *Graph {
	return &Graph{
		nodes: make(map[NodeId]Node),
		edges: make(map[EdgeId]Edge),
	}
}

func (g *Graph) AddNode(node Node) bool{
	if _, exits := g.nodes[node.Id]; exits {
		return false
	} else {
		g.nodes[node.Id] = node
		return true
	}
	return false
}

func (g *Graph) GetNode(nodeId NodeId) Node {
	return g.nodes[nodeId]
}

func (g *Graph) AddEdge(edge Edge) bool{
	if _, exits := g.edges[edge.Id]; exits {
		return false
	} else {
		g.edges[edge.Id] = edge
		return true
	}
	return false
}

func (g *Graph) String() string {
	return fmt.Sprintf("%+v\n %+v", g.nodes, g.edges)
}

type Node struct {
	Id NodeId
}

type Edge struct {
	Id   EdgeId
	Head NodeId
	Tail NodeId
}

File: client/ts/game/graph/Graph.ts
/**
Modified typescript version of https://github.com/davidpiegza/Graph-Visualization
**/
import Edge = require("game/graph/Edge");

class Graph {
    private _nodeSet
    public nodes
    public edges
    
    constructor() {
        this._nodeSet = {};
        this.nodes = [];
        this.edges = []; 
    }
    public addNode(node):boolean{
        if(this._nodeSet[node.id] == undefined) {
            this._nodeSet[node.id] = node;
            this.nodes.push(node);
            return true;
        }
        return false;
    } 
    public getNode(node_id):any {
        return this._nodeSet[node_id];
    }
    public addEdge(edge_id, source, target):boolean{
        if(source.addConnectedTo(target) === true) {
            var edge = new Edge(edge_id, source, target);
            this.edges.push(edge);
            return true;
        }
        return false;
    }

}

export = Graph